roles/v2/mining-proxy/src/lib/upstream_mining.rs

```diff
@@ -110,9 +130,137 @@ impl UpstreamMiningNode {
+                    .safe_lock(|a| {
+                        a.channel_factory
+                            .as_mut()
+                            .unwrap()
+                            .on_new_template(&mut message_new_template)
+                    })
+                    .unwrap()
+                    .unwrap();
+                if let Some(custom_job) = custom_job {
+                    let req_id = self_mutex
+                        .safe_lock(|s| s.request_ids.safe_lock(|r| r.next()).unwrap())
+                        .unwrap();
+                    let custom_mining_job =
+                        PoolMessages::Mining(Mining::SetCustomMiningJob(SetCustomMiningJob {
+                            channel_id: self_mutex.safe_lock(|s| s.id).unwrap(),
+                            request_id: req_id,
+                            version: custom_job.version,
+                            prev_hash: custom_job.prev_hash,
+                            min_ntime: custom_job.min_ntime,
+                            nbits: custom_job.nbits,
+                            coinbase_tx_version: custom_job.coinbase_tx_version,
+                            coinbase_prefix: custom_job.coinbase_prefix.clone(),
+                            coinbase_tx_input_n_sequence: custom_job.coinbase_tx_input_n_sequence,
+                            coinbase_tx_value_remaining: custom_job.coinbase_tx_value_remaining,
+                            coinbase_tx_outputs: custom_job.coinbase_tx_outputs.clone(),
+                            coinbase_tx_locktime: custom_job.coinbase_tx_locktime,
+                            merkle_path: custom_job.merkle_path,
+                            extranonce_size: custom_job.extranonce_size,
+                            future_job: message_new_template.future_template,
+                            token,
+                        }));
+                    Self::send(self_mutex.clone(), custom_mining_job.try_into().unwrap())
+                        .await
+                        .unwrap();
+                }
+
+                for (id, job) in channel_id_to_new_job_msg {
+                    let downstream = self_mutex
+                        .safe_lock(|a| a.downstream_selector.downstream_from_channel_id(id))
+                        .unwrap()
+                        .unwrap();
+                    let message = MiningDeviceMessages::Mining(job);
+                    DownstreamMiningNode::send(downstream, message.try_into().unwrap())
+                        .await
+                        .unwrap();
+                }
+            }
+        });
+
+        //pub fn next_match()
+    }
+
+    pub fn start_receiving_new_prev_hash(self_mutex: Arc<Mutex<Self>>) {
+        task::spawn(async move {
+            while self_mutex
+                .safe_lock(|s| s.channel_factory.is_none())
+                .unwrap()
+            {
+                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
```

Comment:
I've seen this loop a few times? Make a function?

Reply:
Done